grpc/src/main/java/com/linecorp/armeria/internal/grpc/ArmeriaMessageDeframer.java

```diff
@@ -0,0 +1,503 @@
+            * has requested another message.  At this point in the function, either all frames have been
+            * delivered, or unprocessed is empty.  If there is a partial message, it will be inside next
+            * frame and not in unprocessed.  If there is extra data but no pending deliveries, it will
+            * be in unprocessed.
+            */
+            boolean stalled = unprocessed.readableBytes() == 0;
+
+            if (endOfStream && stalled) {
+                boolean havePartialMessage = nextFrame != null && nextFrame.readableBytes() > 0;
+                if (!havePartialMessage) {
+                    listener.endOfStream();
+                    deliveryStalled = false;
+                    return;
+                } else {
+                    // We've received the entire stream and have data available but we don't have
+                    // enough to read the next frame ... this is bad.
+                    throw Status.INTERNAL.withDescription(
+                            DEBUG_STRING + ": Encountered end-of-stream mid-frame").asRuntimeException();
+                }
+            }
+            deliveryStalled = stalled;
+        } finally {
+            inDelivery = false;
+        }
+    }
+
+    /**
+     * Attempts to read the required bytes into nextFrame.
+     *
+     * @return {@code true} if all of the required bytes have been read.
+     */
+    private boolean readRequiredBytes() {
+        if (nextFrame == null) {
+            nextFrame = alloc.compositeBuffer();
+        }
+
+        // Read until the buffer contains all the required bytes.
+        int missingBytes;
+        while ((missingBytes = requiredLength - nextFrame.readableBytes()) > 0) {
+            if (unprocessed.readableBytes() == 0) {
+                // No more data is available.
+                return false;
+            }
+            int toRead = Math.min(missingBytes, unprocessed.readableBytes());
+            nextFrame.addComponent(true, unprocessed.readBytes(toRead));
+        }
+        return true;
+    }
+
+    /**
+     * Processes the GRPC compression header which is composed of the compression flag and the outer
+     * frame length.
+     */
+    private void processHeader() {
+        int type = nextFrame.readUnsignedByte();
+        if ((type & RESERVED_MASK) != 0) {
+            throw Status.INTERNAL.withDescription(
+                    DEBUG_STRING + ": Frame header malformed: reserved bits not zero")
+                                 .asRuntimeException();
+        }
+        compressedFlag = (type & COMPRESSED_FLAG_MASK) != 0;
+
+        // Update the required length to include the length of the frame.
+        requiredLength = nextFrame.readInt();
+        if (requiredLength < 0 || requiredLength > maxMessageSizeBytes) {
+            throw Status.INTERNAL.withDescription(String.format("%s: Frame size %d exceeds maximum: %d. ",
```

Comment:
This error message is straight from grpc-java so let's keep it as is for consistency.